File: src/engine/board.ts
/**
 * กระดานของหมากสี่กระดาน
 *
 * กระดาน 8x8 = 64 ช่อง โดยช่องกลาง 4 ช่องเป็น "ช่องว่างเริ่มต้น"
 * จึงเริ่มเกมด้วยหมาก 60 ตัวบนอีก 60 ช่อง และมีช่องว่างให้กระโดดลงได้ทันที
 *
 * ช่องถูกอ้างอิงด้วย index เดียว: i = r * 8 + c  (r = 0 คือแถวบนสุด)
 */

export const SIZE = 8;
export const CELLS = SIZE * SIZE;

/** ช่องกลาง 4 ช่อง: (3,3) (3,4) (4,3) (4,4) แบบ 0-indexed */
export const CENTER_HOLES: readonly number[] = [27, 28, 35, 36];

export const PIECE_COUNT = CELLS - CENTER_HOLES.length; // 60

/** null = ช่องว่าง, number = piece id ของหมากที่อยู่บนช่องนั้น */
export type Board = (number | null)[];

export interface Dir {
  readonly dr: number;
  readonly dc: number;
  /** สัญลักษณ์ไว้ใช้ใน log / debug */
  readonly glyph: string;
}

/** 8 ทิศทาง: แนว + และแนว X */
export const DIRS: readonly Dir[] = [
  { dr: -1, dc: 0, glyph: "↑" },
  { dr: 1, dc: 0, glyph: "↓" },
  { dr: 0, dc: -1, glyph: "←" },
  { dr: 0, dc: 1, glyph: "→" },
  { dr: -1, dc: -1, glyph: "↖" },
  { dr: -1, dc: 1, glyph: "↗" },
  { dr: 1, dc: -1, glyph: "↙" },
  { dr: 1, dc: 1, glyph: "↘" },
];

export const idx = (r: number, c: number): number => r * SIZE + c;
export const rowOf = (i: number): number => (i / SIZE) | 0;
export const colOf = (i: number): number => i % SIZE;

/**
 * เดินจากช่อง i ไปตามทิศ d เป็นจำนวน n ก้าว
 * คืน -1 ถ้าหลุดขอบกระดาน
 */
export function shift(i: number, d: Dir, n = 1): number {
  const r = rowOf(i) + d.dr * n;
  const c = colOf(i) + d.dc * n;
  if (r < 0 || r >= SIZE || c < 0 || c >= SIZE) return -1;
  return idx(r, c);
}

/** กระดานเริ่มต้น: หมาก 60 ตัว (id 0..59) เว้นช่องกลาง 4 ช่อง */
export function initialBoard(): Board {
  const holes = new Set(CENTER_HOLES);
  const board: Board = new Array(CELLS).fill(null);
  let id = 0;
  for (let i = 0; i < CELLS; i++) {
    if (!holes.has(i)) board[i] = id++;
  }
  return board;
}

/** ตำแหน่งแบบอ่านง่าย เช่น 12 -> "e7" */
export function squareLabel(i: number): string {
  return `${"abcdefgh"[colOf(i)]}${SIZE - rowOf(i)}`;
}

export function findPiece(board: Board, pieceId: number): number {
  return board.indexOf(pieceId);
}

export function countPieces(board: Board): number {
  let n = 0;
  for (const cell of board) if (cell !== null) n++;
  return n;
}

File: apps/api/src/makthai/games/mak_si_kradan/board.py
"""กระดาน 8x8 ที่มีช่องกลาง 4 ช่องว่างตั้งแต่เริ่ม

ช่องกลางต้องว่างตอนเริ่ม ไม่ใช่ห้ามลงถาวร ไม่งั้นกระดานจะเต็มตั้งแต่เทิร์นแรก
และไม่มีช่องให้กระโดดลงเลย เกมจะไม่มีการเล่นที่ถูกต้องตั้งแต่ต้น
"""

from __future__ import annotations

from dataclasses import dataclass

SIZE = 8
CELLS = SIZE * SIZE

#: ช่องกลาง 4 ช่อง — (3,3) (3,4) (4,3) (4,4) แบบนับจาก 0
CENTER_HOLES: tuple[int, ...] = (27, 28, 35, 36)

PIECE_COUNT = CELLS - len(CENTER_HOLES)

#: None = ช่องว่าง, int = id ของหมากที่อยู่บนช่องนั้น
Board = list[int | None]

FILES = "abcdefgh"


@dataclass(frozen=True, slots=True)
class Direction:
    dr: int
    dc: int
    glyph: str


#: 8 ทิศ — ทั้งแนวตั้ง แนวนอน และแนวทแยง
DIRECTIONS: tuple[Direction, ...] = (
    Direction(-1, 0, "↑"),
    Direction(1, 0, "↓"),
    Direction(0, -1, "←"),
    Direction(0, 1, "→"),
    Direction(-1, -1, "↖"),
    Direction(-1, 1, "↗"),
    Direction(1, -1, "↙"),
    Direction(1, 1, "↘"),
)


def index_of(row: int, col: int) -> int:
    return row * SIZE + col


def row_of(square: int) -> int:
    return square // SIZE


def col_of(square: int) -> int:
    return square % SIZE


def shift(square: int, direction: Direction, steps: int = 1) -> int:
    """เดินจากช่องหนึ่งไปตามทิศ คืน -1 เมื่อหลุดขอบกระดาน"""
    row = row_of(square) + direction.dr * steps
    col = col_of(square) + direction.dc * steps
    if not (0 <= row < SIZE and 0 <= col < SIZE):
        return -1
    return index_of(row, col)


def jumped_square(origin: int, landing: int) -> int:
    """ช่องที่ถูกข้ามระหว่างกระโดดจากช่องหนึ่งไปอีกช่อง"""
    return index_of(
        (row_of(origin) + row_of(landing)) // 2,
        (col_of(origin) + col_of(landing)) // 2,
    )


def initial_board() -> Board:
    """หมาก 60 ตัว เว้นช่องกลาง 4 ช่อง"""
    holes = set(CENTER_HOLES)
    board: Board = [None] * CELLS
    piece_id = 0
    for square in range(CELLS):
        if square not in holes:
            board[square] = piece_id
            piece_id += 1
    return board


def square_label(square: int) -> str:
    """ตำแหน่งแบบอ่านง่าย เช่น 12 -> 'e7'"""
    return f"{FILES[col_of(square)]}{SIZE - row_of(square)}"


def count_pieces(board: Board) -> int:
    return sum(1 for cell in board if cell is not None)


def find_piece(board: Board, piece_id: int) -> int:
    return board.index(piece_id) if piece_id in board else -1
